### PR TITLE
[GLUTEN-8115][CORE] Refine the BuildSideRelation transform to support all scenarios

### PR DESCRIPTION
**Pull request by yikf**
## What changes were proposed in this pull request?
Fix https://github.com/apache/incubator-gluten/issues/8115.

If exchange reuse, the `buildKeys` of `ColumnarSubqueryBroadcastExec` and the `buildKeys` of `BroadcastExchangeExec` may not be the same, making it difficult to construct the projection of the transform. therefore, the `BuildSideRelation` should retain the `BroadcastMode` of `BroadcastExchangeExec` to simplify the projection of the transform.

## How was this patch tested?
exists unit tests, manual tests

### CODE REVIEW

**Comment by yikf** [comment]
@PHILO-HE @zhztheplayer @jackylee-ch @zzcclp Could you please take a look if you find a time, thanks

**Comment by FelixYBW** [comment]
Do you have a specific query example?

**Comment by yikf** [comment]
> Do you have a specific query example?

TPC-DS Q5 is a reuse-exchange example, see: https://github.com/apache/incubator-gluten/issues/7807

After reuse, ColumnarSubqueryBroadcastExec's buildKeys and exchange's buildKeys may not be the same reference, and in our internal case, even the names referenced by alias are different because of alias. 

In Spark, the transform of relation is [always carried out in BroadcastExchangeExec](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/BroadcastExchangeExec.scala#L193), so this problem does not exist. we can save the buildKeys of BroadcastExchangeExec. this way, in the subsequent transform, we can do the relation transform like spark would have done in BroadcastExchangeExec, in which case the references are always the same, and then we can take the desired index value from the buildKeys output.

**Comment by yikf** [comment]
@FelixYBW @zhouyuan @PHILO-HE Hi, Could you help please take a look, thanks.

File: gluten-substrait/src/main/scala/org/apache/spark/sql/execution/ColumnarSubqueryBroadcastExec.scala [hunk 1]
```diff
@@ -64,6 +65,14 @@ case class ColumnarSubqueryBroadcastExec(
     copy(name = "native-dpp", buildKeys = keys, child = child.canonicalized)
   }
 
+  // Copy from org.apache.spark.sql.execution.joins.HashJoin#canRewriteAsLongType
+  // we should keep consistent with it to identify the LongHashRelation.
+  private def canRewriteAsLongType(keys: Seq[Expression]): Boolean = {
```
Comment (PHILO-HE) [inline gluten-substrait/src/main/scala/org/apache/spark/sql/execution/ColumnarSubqueryBroadcastExec.scala:70]: @yikf, if spark changes this logic, e.g., for adding support for other types, unexpected failure can happen in Gluten? Seems there is no other way.

**Review by PHILO-HE — APPROVED** [review]
Verified with internal TPC-DS 3TB test.

File: gluten-substrait/src/main/scala/org/apache/spark/sql/execution/ColumnarSubqueryBroadcastExec.scala (hunk 1 above)
Comment (yikf) [reply gluten-substrait/src/main/scala/org/apache/spark/sql/execution/ColumnarSubqueryBroadcastExec.scala:70]: There doesn't seem to be any other way at the moment.

Comment (PHILO-HE) [reply gluten-substrait/src/main/scala/org/apache/spark/sql/execution/ColumnarSubqueryBroadcastExec.scala:70]: @yikf, ok, let's merge this pr firstly. Thanks for your efforts!

Comment (yikf) [reply gluten-substrait/src/main/scala/org/apache/spark/sql/execution/ColumnarSubqueryBroadcastExec.scala:70]: thanks @PHILO-HE

**Pull request merged**